### PR TITLE
Use routes instead of port-forwarding where possible to improve test …

### PR DESCRIPTION
**Pull request by kevinearls**
…reliability

Signed-off-by: Kevin Earls <kearls@redhat.com>

We get port forwarding failures regularly when running CI on OpenShift.  This updates some tests to use routes where available instead in order to improve test reliability.

### CODE REVIEW

**Comment by objectiser** [comment]
@kevinearls From the description you mention about the failures on OpenShift, but wouldn't the same be true for K8s? Would it not also be worth using the ingress on k8s?

File: test/e2e/smoketest.go [hunk 1]
```diff
@@ -58,10 +58,11 @@ func productionSmokeTest(jaegerInstanceName, smokeTestNamespace string) {
 	queryPodPrefix := jaegerInstanceName + "-query"
 	collectorPodPrefix := jaegerInstanceName + "-collector"
 
+	// Use ingress for k8s or on OpenShift if we have an insecure route
 	var apiTracesEndpoint string
-	route, hasInsecureRoute := getInsecureRoute(jaegerInstanceName)
-	if hasInsecureRoute {
-		apiTracesEndpoint = fmt.Sprintf("https://%s/api/traces", route.Spec.Host)
+	hasInsecureRoute := hasInsecureRoute(jaegerInstanceName)
+	if !isOpenShift(t) || hasInsecureRoute {
```
Comment (objectiser) [inline test/e2e/smoketest.go]: Would be better to move the test of openshift into the `hasInsecureRoute` function, and then this condition could just be `if hasInsecureRoute {`.

Comment (objectiser) [reply test/e2e/smoketest.go]: Also, then maybe change the method name, so doesn't imply specific to route - e.g. `hasInsecureEndpoint`?

**Comment by objectiser** [comment]
@kevinearls  Sorry approved to soon - can be merged subject to CI passing.

**Comment by kevinearls** [comment]
@jpkrohling @pavolloffay  Please review.  @objectiser  please re-review.

File: test/e2e/cassandra_test.go [hunk 2]
```diff
@@ -85,17 +88,27 @@ func (suite *CassandraTestSuite) TestCassandraSparkDependencies() {
...
 		},
 		ObjectMeta: metav1.ObjectMeta{
-			Name:      "with-cassandra",
+			Name:      jaegerInstanceName,
 			Namespace: namespace,
 		},
 		Spec: v1.JaegerSpec{
+			Ingress: v1.JaegerIngressSpec{
+				Enabled:  &ingressEnabled,
+				Security: v1.IngressSecurityNoneExplicit,
```
Comment (jpkrohling) [inline test/e2e/cassandra_test.go]: Why is this needed? For OpenShift? We should really try to get the OpenShift service-ca trusted by the client. If it's not feasible for this PR, please create an issue to track it.

File: test/e2e/smoketest.go [hunk 3]
```diff
@@ -83,9 +114,17 @@ func executeSmokeTest(apiTracesEndpoint, collectorEndpoint string) {
 		Finish()
 	closer.Close()
 
+	transport := &http.Transport{}
+	if hasInsecureEndpoint {
+		insecure := true
+		// #nosec  G402: TLS InsecureSkipVerify set true
+		transport.TLSClientConfig = &tls.Config{InsecureSkipVerify: insecure}
```
Comment (jpkrohling) [inline test/e2e/smoketest.go]: Same comment as before: we should try to get the service-ca trusted by the client instead of skipping the cert verification.

Comment (kevinearls) [reply test/e2e/smoketest.go]: Agreed.  Unfortunately I don't really know how to do this, and my initial attempts to get something working have failed.  So any tips or pointers to appropriate doc would be appreciated.

Comment (jpkrohling) [reply test/e2e/smoketest.go]: There are a couple of options to get the `service-ca`: 

1. if your test is running inside the cluster (I don't think it is), you can just get the file at `/var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt ` and use that as the CA.
1. if the test is outside of the cluster, you can create a secret annotated with `service.beta.openshift.io/inject-cabundle=true`, and the CA bundle will be at the key `service-ca.crt`.
1. you can also extract the bundle from the existing operator deployment, using `kubectl cp`. But this should be the last resort.

Options 1 and 2 are described here: https://docs.openshift.com/container-platform/4.3/authentication/certificates/service-serving-certificate.html

Comment (objectiser) [reply test/e2e/smoketest.go]: @jpkrohling Wondering if this should be tackled in a separate PR? So initially see whether this approach of moving away from port forwarding (as much as possible) improves the reliability of the tests?

Comment (jpkrohling) [reply test/e2e/smoketest.go]: Absolutely, as long as there's an issue to track this

Comment (objectiser) [reply test/e2e/smoketest.go]: Ok great, on that basis are you ok with the PR as is?

Comment (jpkrohling) [reply test/e2e/smoketest.go]: LGTM
